src/java.base/share/classes/java/util/TreeMap.java

```diff
@@ -3322,16 +3314,8 @@ public int characteristics() {
         @Override
         public Comparator<Map.Entry<K, V>> getComparator() {
             // Adapt or create a key-based comparator
-            if (tree.comparator != null) {
-                return Map.Entry.comparingByKey(tree.comparator);
-            }
-            else {
-                return (Comparator<Map.Entry<K, V>> & Serializable) (e1, e2) -> {
-                    @SuppressWarnings("unchecked")
-                    Comparable<? super K> k1 = (Comparable<? super K>) e1.getKey();
-                    return k1.compareTo(e2.getKey());
-                };
-            }
+            Comparator<? super K> treeComparator = tree.comparator;
+            return Map.Entry.comparingByKey(treeComparator == null ? natural() : treeComparator);
```

Comment:
You can probably have:
```java
return treeComparator == nul ?
        Map.Entry.comparingByKey() :
        Map.Entry.comparingByKey(treeComparator);
```
instead.

Reply:
Nope, there'd be a compilation error

Reply:
It'll work fine if an unchecked cast is used:
```suggestion
            @SuppressWarnings("unchecked")
            Comparator<Map.Entry<K, V>> entryComparator = (Comparator) (treeComparator == null ?
                    Map.Entry.comparingByKey()
                    Map.Entry.comparingByKey(treeComparator));
            return entryComparator;
```

This also ensures that when `treeComparator` is `null`, this method keeps returning a constant lambda.

Reply:
This one looks fine, accepted.